cgi-bin/DW/Controller/Admin/SendMail.pm

```diff
@@ -171,5 +171,122 @@ sub index_controller {
+        }
+        # sendto might be a uid
+        if ( $row->{sendto} =~ /^\d+$/ ) {
+            if ( my $u = LJ::load_userid( $row->{sendto} ) ) {
+                $row->{sendto} = $u;
+            } else {
+                $row->{sendto} = '[unknown user] '. $row->{sendto};
+            }
+        }
+        # sentfrom is definitely a uid
+        if ( my $ru = LJ::load_userid( $row->{remoteid} ) ) {
+            $row->{remote} = $ru;
+        } else {
+            $row->{remote} = '[unknown user] '. $row->{remoteid};
+        }
+        # add domain name to display site address
+        $row->{account_view} = $row->{account} . "\@$LJ::DOMAIN";
+
+        return $row;
+    };
+
+    my $r = DW::Request->get;
+    my $errors = DW::FormErrors->new;
+
+    if ( $r->did_post ) {
+        my $args = $r->post_args;
+
+        my $account = LJ::trim( $args->{account} );
+
+        if ( $account ) {
+            if ( my $priv = $LJ::SENDMAIL_ACCOUNTS{ $account } ) {
+                $errors->add( "account", ".error.nopriv", { priv => $priv } )
+                    unless $remote->has_priv( $priv );
+            } else {
+                $errors->add( "account", ".error.badacct" );
+            }
+
+        } else {
+            $errors->add( "account", ".error.noacct" );
+        }
+
+        unless ( $errors->exist ) {
+            my $dbr = LJ::get_db_reader();
+            my $rows = $dbr->selectall_hashref(
+                "SELECT msgid, time_sent, sendto, subject, request" .
+                " FROM siteadmin_email_history WHERE account=?", 'msgid',
+                undef, $account );
+            die $dbr->errstr if $dbr->err;
+
+            my @data = map { $enhance_row->( $_ ) } values %$rows;
+
+            $rv->{rows} = [ sort { $b->{time_sent} <=> $a->{time_sent} } @data ];
+            $rv->{account}  = $account;
+        }
+
+    } else {
+        my $args = $r->get_args;
+
+        if ( my $msgid = LJ::trim( $args->{id} ) ) {
+            return $r->redirect( "/admin/sendmail/" )
+                unless $msgid =~ /^\d+$/;
+
+            my $dbr = LJ::get_db_reader();
+            my $row = $dbr->selectrow_hashref(
+                "SELECT * FROM siteadmin_email_history WHERE msgid=?",
+                undef, $msgid );
+            die $dbr->errstr if $dbr->err;
+
+            return error_ml( $scope->( '.error.nomsg' ) ) unless $row;
+            $row = $enhance_row->( $row );
+
+            my $priv = $LJ::SENDMAIL_ACCOUNTS{ $row->{account} };
+
+            if ( $priv && $remote->has_priv( $priv ) ) {
+                $rv->{row} = $row;
+            } else {
+                return error_ml( $scope->( '.error.nopriv' ), { priv => $priv } );
+            }
+
+            return DW::Template->render_template( 'admin/sendmail/message.tt', $rv );
+        }
+    }
+
+    # Construct data for dropdown of available email addresses
+    my @account_menu = ( "", LJ::Lang::ml( $scope->( '.select.account.choose' ) ) );
```

Comment:
I feel like I saw this last iteration as well. Possible candidate for moving into its own method to reduce duplication.